### PR TITLE
Spirv_intrinsics: Add support of type specifier to spirv_type

### PR DESCRIPTION
**Pull request by amdrexu**
Previously, spirv_type doesn't accept type specifier as its parameter. With this change, we can input non-array type specifier. This is because some SPIR-V type definition intructions often need to reference other SPIR-V types as its source operands. We add the support to facilitate such usage.

### CODE REVIEW

File: glslang/Include/SpirvIntrinsics.h [hunk 1]
```diff
@@ -98,12 +98,23 @@ struct TSpirvInstruction {
...
+        type = arg;
+    }
 
-    bool operator==(const TSpirvTypeParameter& rhs) const { return constant == rhs.constant; }
+    bool operator==(const TSpirvTypeParameter& rhs) const;
     bool operator!=(const TSpirvTypeParameter& rhs) const { return !operator==(rhs); }
 
-    const TIntermConstantUnion* constant;
+    const TIntermConstantUnion* constant; // Constant expression
+    const TType* type;                    // Type specifier
```
Comment (arcady-lunarg) [inline glslang/Include/SpirvIntrinsics.h:117]: This feels like an opportunity to use something like std::variant from C++17, now that we require C++17 as a minimum, but I can do that refactor in a followup change.

File: glslang/Include/SpirvIntrinsics.h [hunk 2]
```diff
@@ -65,19 +65,19 @@ struct TSpirvExecutionMode {
     // spirv_execution_mode
     TMap<int, TVector<const TIntermConstantUnion*>> modes;
     // spirv_execution_mode_id
-    TMap<int, TVector<const TIntermTyped*> > modeIds;
+    TMap<int, TVector<const TIntermTyped*>> modeIds;
```
Comment (arcady-lunarg) [inline glslang/Include/SpirvIntrinsics.h:68]: Could you remove these unrelated whitespace diffs?

File: glslang/Include/SpirvIntrinsics.h [hunk 3]
```diff
@@ -65,19 +65,19 @@ struct TSpirvExecutionMode {
...
 };
 
 // SPIR-V decorations
 struct TSpirvDecorate {
     POOL_ALLOCATOR_NEW_DELETE(GetThreadPoolAllocator())
 
     // spirv_decorate
-    TMap<int, TVector<const TIntermConstantUnion*> > decorates;
+    TMap<int, TVector<const TIntermConstantUnion*>> decorates;
```
Comment (arcady-lunarg) [inline glslang/Include/SpirvIntrinsics.h:76]: Unrelated whitespace diff.

File: glslang/Include/SpirvIntrinsics.h [hunk 4]
```diff
@@ -65,19 +65,19 @@ struct TSpirvExecutionMode {
...
     // spirv_decorate_id
     TMap<int, TVector<const TIntermTyped*>> decorateIds;
     // spirv_decorate_string
-    TMap<int, TVector<const TIntermConstantUnion*> > decorateStrings;
+    TMap<int, TVector<const TIntermConstantUnion*>> decorateStrings;
```
Comment (arcady-lunarg) [inline glslang/Include/SpirvIntrinsics.h:80]: Unrelated whitespace diff.

File: glslang/MachineIndependent/SpirvIntrinsics.cpp [hunk 5]
```diff
@@ -274,23 +283,28 @@ void TPublicType::setSpirvType(const TSpirvInstruction& spirvInst, const TSpirvT
...
-        constant->getBasicType() != EbtInt &&
-        constant->getBasicType() != EbtUint &&
-        constant->getBasicType() != EbtBool &&
+
+    if (constant->getBasicType() != EbtFloat && constant->getBasicType() != EbtInt &&
+        constant->getBasicType() != EbtUint && constant->getBasicType() != EbtBool &&
         constant->getBasicType() != EbtString)
         error(loc, "this type not allowed", constant->getType().getBasicString(), "");
-    else {
-        assert(constant);
```
Comment (arcady-lunarg) [inline glslang/MachineIndependent/SpirvIntrinsics.cpp:287]: This assert is already useless due to the if's condition dereferencing constant. Might it make sense to move it up to the top of the function instead?

File: glslang/MachineIndependent/SpirvIntrinsics.cpp [hunk 6]
```diff
@@ -274,23 +283,28 @@ void TPublicType::setSpirvType(const TSpirvInstruction& spirvInst, const TSpirvT
         spirvType->typeParams = *typeParams;
 }
 
-TSpirvTypeParameters* TParseContext::makeSpirvTypeParameters(const TSourceLoc& loc, const TIntermConstantUnion* constant)
+TSpirvTypeParameters* TParseContext::makeSpirvTypeParameters(const TSourceLoc& loc,
+                                                             const TIntermConstantUnion* constant)
```
Comment (arcady-lunarg) [inline glslang/MachineIndependent/SpirvIntrinsics.cpp:287]: Unrelated whitespace change

File: glslang/MachineIndependent/SpirvIntrinsics.cpp [hunk 7]
```diff
@@ -274,23 +283,28 @@ void TPublicType::setSpirvType(const TSpirvInstruction& spirvInst, const TSpirvT
...
 {
     TSpirvTypeParameters* spirvTypeParams = new TSpirvTypeParameters;
-    if (constant->getBasicType() != EbtFloat &&
-        constant->getBasicType() != EbtInt &&
-        constant->getBasicType() != EbtUint &&
-        constant->getBasicType() != EbtBool &&
+
+    if (constant->getBasicType() != EbtFloat && constant->getBasicType() != EbtInt &&
```
Comment (arcady-lunarg) [inline glslang/MachineIndependent/SpirvIntrinsics.cpp:291]: Unrelated whitespace diff.

**Review by arcady-lunarg — CHANGES_REQUESTED** [review]
On the whole this is fine other than some unrelated whitespace changes that I would prefer you remove. While it would be nice to use std::variant here, I am fine doing that myself in a follow-up change.

**Review by arcady-lunarg — APPROVED** [review]
I did the whitespace fixes myself and updated your branch.

**Comment by amdrexu** [comment]
> I did the whitespace fixes myself and updated your branch.

Thank you for your kind help. I take leaves and didn't notice the review comments :)

**Pull request merged**